torchtext/datasets/language_modeling.py

```diff
@@ -1,217 +1,199 @@
-            Remaining keyword arguments: Passed to the splits method.
         """
-        TEXT = data.Field()
-
-        train, val, test = cls.splits(TEXT, root=root, **kwargs)
-
-        TEXT.build_vocab(train, vectors=vectors)
-
-        return data.BPTTIterator.splits(
-            (train, val, test), batch_size=batch_size, bptt_len=bptt_len,
-            device=device)
 
+        super(LanguageModelingDataset, self).__init__()
+        self._data = data
+        self._vocab = vocab
+
+    def __getitem__(self, i):
+        return self._data[i]
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        for x in self._data:
+            yield x
+
+    def get_vocab(self):
+        return self._vocab
+
+
+def _setup_datasets(dataset_name, tokenizer=get_tokenizer("basic_english"),
+                    root='.data', vocab=None, removed_tokens=['<unk>']):
+    if dataset_name == 'PennTreebank':
+        train_path = download_from_url(URLS['PennTreebank'][0], root=root)
+        test_path = download_from_url(URLS['PennTreebank'][1], root=root)
+        valid_path = download_from_url(URLS['PennTreebank'][2], root=root)
+    else:
+        dataset_tar = download_from_url(URLS[dataset_name], root=root)
+        extracted_files = extract_archive(dataset_tar)
+        for fname in extracted_files:
+            if 'train' in fname:
+                train_path = os.path.join(root, fname)
+            if 'test' in fname:
+                test_path = os.path.join(root, fname)
+            if 'valid' in fname:
+                valid_path = os.path.join(root, fname)
+
+    if vocab is None:
+        logging.info('Building Vocab based on {}'.format(train_path))
+        vocab = build_vocab_from_iterator(read_text_iterator(train_path, tokenizer))
+    else:
+        if not isinstance(vocab, Vocab):
+            raise TypeError("Passed vocabulary is not of type Vocab")
+    logging.info('Vocab has {} entries'.format(len(vocab)))
+    logging.info('Creating training data')
+    train_iter = create_data_from_iterator(
+        vocab, read_text_iterator(train_path, tokenizer), removed_tokens)
+    train_data = []
+    for tokens in train_iter:
+        train_data += tokens
+
+    logging.info('Creating testing data')
+    test_iter = create_data_from_iterator(
+        vocab, read_text_iterator(test_path, tokenizer), removed_tokens)
+    test_data = []
+    for tokens in test_iter:
+        test_data += tokens
+
+    logging.info('Creating valid data')
+    valid_iter = create_data_from_iterator(
+        vocab, read_text_iterator(valid_path, tokenizer), removed_tokens)
+    valid_data = []
+    for tokens in valid_iter:
+        valid_data += tokens
+
+    return (LanguageModelingDataset(torch.Tensor(train_data).long(), vocab),
+            LanguageModelingDataset(torch.Tensor(test_data).long(), vocab),
+            LanguageModelingDataset(torch.Tensor(valid_data).long(), vocab))
```

Comment:
A "dataset" is therefore a list of 3 datasets corresponding to train/test/validation? I assume the division is dictated by the source itself, right?

Could the user ever want to just get one of train/test/validation? What do you think of letting the user get one at the time when building a dataset?

Reply:
For reference: [librispeech](https://github.com/pytorch/audio/blob/38d1a9b6325d8b69e9b26ce28b1f1a8e0040c51b/torchaudio/datasets/librispeech.py#L63) in audio simply lets the user chose which one to download.

Reply:
That could be an option with an extra input for that.
But it's kind of convention to generate three datasets together (if there are) and most users will need all the them for training and inference.

Reply:
Would it be more flexible to have a function returning each separately, and then wrap all three with the API here?

Something like 
```
train = _setup_datasets_part(train)
test = _setup_datasets_part(test)
valid = _setup_datasets_part(valid)
return train, test, valid
```

Reply:
the flag is probably a good idea. Someone might only want to run validation or testing in a separate process. either accept a string or a tuple of strings i'd say.

Reply:
Is this consistent with the supervised datasets? If not, maybe we should add that option.

Reply:
Supervised datasets return train/test datasets all together. I feel that we should keep consistent among DAPIs or at least within torchtext. Of course, an extra input flag to select the dataset adds more flexibility :)

One usage case is concerning me: say you want valid dataset only (as proposed), then you have to call the API to generate train dataset again because the vocab object is generated based on train data. Instead, if we generate test/train/valid all together, then we re-use the vocab object for valid/test datasets.

Reply:
I'd say, by default, we generate all three, but then we allow the user to create a single dataset. For that it's still up to the user whether they want to pass a custom (pretrained) Vocab, but we can mention in the docs that they probably want in most cases.

I'd expect a common usecase to be that a user generated all three in a training script and then only the test data in a separate evaluation script / context and wants to pass a Vocab. Someone might also only want to load the test dataset as yet another evaluation opportunity after having trained on another dataset (e.g. train on wikitext2, evaluate on penntreebank using a sentencepiece vocab).

On that note, if I want to inject sentencepiece as a vocab here, does that work?

Reply:
> Instead, if we generate test/train/valid all together, then we re-use the vocab object for valid/test datasets.

What I would expect here is to have a function that produce one of the datasets given a vocab, similar to `_setup_datasets` but for only one of the datasets. We can then produce the vocab once, and pass it along when we create each datasets. Thoughts?

This can then be wrapped in a function like `_setup_datasets` which then returns all three together. This preserves consistency in torchtext, but also allows only one of the three to be generated, say as an evaluation script as mentioned by @cpuhrsch. 

> On that note, if I want to inject sentencepiece as a vocab here, does that work?

The API suggests, so: `LanguageModelingDataset(torch.Tensor(train_data).long(), vocab)` -- @zhangguanheng66 ?

Reply:
> I'd say, by default, we generate all three, but then we allow the user to create a single dataset. For that it's still up to the user whether they want to pass a custom (pretrained) Vocab, but we can mention in the docs that they probably want in most cases.
> 
> I'd expect a common usecase to be that a user generated all three in a training script and then only the test data in a separate evaluation script / context and wants to pass a Vocab. Someone might also only want to load the test dataset as yet another evaluation opportunity after having trained on another dataset (e.g. train on wikitext2, evaluate on penntreebank using a sentencepiece vocab).
> 
> On that note, if I want to inject sentencepiece as a vocab here, does that work?

Yes. As long as the `vocab` object is a callable function to convert tokens into ids.

In general, an option to generate a single dataset is not hard. I will think about the API.

Reply:
> > Instead, if we generate test/train/valid all together, then we re-use the vocab object for valid/test datasets.
> 
> What I would expect here is to have a function that produce one of the datasets given a vocab, similar to `_setup_datasets` but for only one of the datasets. We can then produce the vocab once, and pass it along when we create each datasets. Thoughts?
> 
> This can then be wrapped in a function like `_setup_datasets` which then returns all three together. This preserves consistency in torchtext, but also allows only one of the three to be generated, say as an evaluation script as mentioned by @cpuhrsch.
> 
> > On that note, if I want to inject sentencepiece as a vocab here, does that work?
> 
> The API suggests, so: `LanguageModelingDataset(torch.Tensor(train_data).long(), vocab)` -- @zhangguanheng66 ?

It's kind of the current API for `LanguageModelingDataset`. I'm thinking about some changes in `WikiText2`.

Reply:
I'm thinking about adding three input flags `train_filename`, `valid_filename`, `test_filename` in `WikiText2`. By default, those three flags have the corresponding file names.

Those input flags are necessary for other datasets, like translation datasets. For example, under German to English category, there are several training raw data files which represents different years (WMT_2009, WMT_2010, WMT_2011, etc). In this case, users have to choose which one they want to load.

Same thing here. By default, users will have train/valid/test datasets. However, if they turn any of `train_filename`, `valid_filename`, `test_filename` flags to `None`, the code won't generate the corresponding datasets. 
@cpuhrsch @vincentqb

Reply:
We talked about this offline. The decision is to add a keyword argument (a tuple of strings) to identify which datasets will be generated. The default value is a tuple of ('train', 'test'). Users have the flexibility to choose a single dataset with a proper `vocab` object.

Reply:
Ok. 

In [CommonVoice](https://github.com/pytorch/audio/blob/38d1a9b6325d8b69e9b26ce28b1f1a8e0040c51b/torchaudio/datasets/commonvoice.py#L8), the implementation doesn't know about the meaning of train/test/etc. Instead, the zip file contains many tsv files (e.g. for train, test, ...), and the user simply specifies which file to use. 

An advantage is that the user could decide to create its own custom tsv file (say for training), and load that one.

Reply:
> Ok.
> 
> In [CommonVoice](https://github.com/pytorch/audio/blob/38d1a9b6325d8b69e9b26ce28b1f1a8e0040c51b/torchaudio/datasets/commonvoice.py#L8), the implementation doesn't know about the meaning of train/test/etc. Instead, the zip file contains many tsv files (e.g. for train, test, ...), and the user simply specifies which file to use.
> 
> An advantage is that the user could decide to create its own custom tsv file (say for training), and load that one.

Yes, that's kind of the same case for translation dataset so we need to provide an option to choose the file.
However, for most text dataset, the raw data files have been properly tagged with `train`, `test`, and `valid`.

Reply:
Some datasets in audio even have "train_100", "train_500" etc built-in, see [this](https://github.com/pytorch/audio/blob/38d1a9b6325d8b69e9b26ce28b1f1a8e0040c51b/torchaudio/datasets/librispeech.py#L67). With only "train", "test" "valid", how would you like to deal with that here?

Reply:
In my early response, I propose to add three more flags `train_filename`, `valid_filename`, `test_filename`, which give you the flexibility to choose a specific file for training. I think either way should way and we just need to keep consistent across datasets within domain.